shared/timeutils: Standardize supported date range on all platforms.

yoctopuce opened the pull request:
This is pull request is to ensure that time functions (`mktime`, `localtime` and `gmtime`) work properly on a reasonable date range, on all platforms, regardless of the epoch. 

It does fix issue #17340

### Summary

The most important point is to have a test case that checks conversion from timestamp to tuple and back for every day in the _supported time range_. This was currently missing. The new test code starts from 2001, first iterates *backward* on tuple to find the earliest supported date and then goes *forward* by adding 86400 to the timestamp, to verify that the date computations are correct, including leap years, and do not raise an overflow.

The second part of the pull request is to ensure that this test does actually pass on all platforms. 

The most severely affected platforms were 32 bit platforms using a 1970 epoch, as

1. there was an assumption that timestamps can be stored in a `mp_int_t`
2. conversion from the 1970 epoch to the 2000 epoch used internally by timeutils breaks computations for dates before 2000

On some of these platforms, the resulting supported date range was only 2001 to 2037. As 2038 is now only 13 years away, this clearly has to be fixed quickly.

Small code footprint beeing a key value for MicroPython, the suggested minimal _supported time range_ for all platforms is 1970 to 2099. This range can be covered using 32 bit timestamps on ressource-limited platforms, fixes the 1970 epoch problems and requires less code for handling of leap years.

Support for years 2100 and beyond is only enabled for platforms using 64 bit pointers, which are typically less ressource-constrained.


### Testing

The PR includes new test cases that verify the supported time range using architecture-independent Python code.


### Trade-offs and Alternatives

The new `timeutils` code has a smaller footprint than the original for 32 bit machines, as it does not have to handle signed inputs and does not have to handle century-based leap years. 

We could have opted to use signed timestamps centered on 2000, as was probably the original intent of the timeutils code. We felt however that it was more useful to fully support the 21st century rather going supporting dates back to 1940's but breaking in 2068. Handling timestamps as unsigned also simplifies the code by removing some checks for negative values.

We could have opted to upgrade all platforms to use 64-bit signed integers for time functions, which would have offered the broadest possible range for abstract date operations, spanning from the year 1600 to the year 3000 and beyond. However, this would have had a code size impact on 32-bit platforms, with little added value. We guessed that, if a specific application on a 32-bit platform required handling _exotic_ dates, a Python-based implementation would remain a good option.

Josverl commented:
Good to have the tests, and the improvements and consistency across ports.

Does that now also make the tuple order and sizes across all ports the same?

yoctopuce commented:
I wasn't even aware that some ports did not follow the documentation for the tuple content. Given how the test code is written, my guess is that if the order is not the expected one for year, month and day the test will certainly fail. The code does not check the content of time, this could be added, at least to ensure some level of consistency (tuple order).

As for the extra numbers and tuple size 8/9, this was not in the scope of this PR. I was considering submitting a separate PR to provide an optional support for a simple TZ strings with DST support, so that devices can provide proper localtime support for logs for instance. But I first have to implement it :-)

dpgeorge reviewed (commented):
Thanks for this PR.  This is something that's been on my to-do list for a while, so it's great to see it finally get fixed properly.

What's required to switch bare-metal ports to use 64-bit range for timestamps?

yoctopuce commented on ports/esp8266/modtime.c:
```diff
@@ -50,5 +50,5 @@ static mp_obj_t mp_time_localtime_get(void) {
 static mp_obj_t mp_time_time_get(void) {
     // get date and time
-    return mp_obj_new_int(pyb_rtc_get_us_since_epoch() / 1000 / 1000);
+    return mp_obj_new_int_from_uint((mp_uint_t)(pyb_rtc_get_us_since_epoch() / 1000u / 1000u));
```
Some C compilers issue a warning when assigning 64bit int into 32bit without explicit cast, so I tend to always add this cast. But obviously this is not the case for this port, as the same expression a few lines above does not cause a warning, so I will revert this change.

yoctopuce commented:
> Thanks for this PR. This is something that's been on my to-do list for a while, so it's great to see it finally get fixed properly.
> 
> What's required to switch bare-metal ports to use 64-bit range for timestamps?

Any bare-metal port with an epoch at 2000 would actually need a 64 signed timestamp to properly encode time on the whole range.

Internally, timeutils can still use the unsigned 32 bit representation to reduce code footprint, but the port should be using 64 bit signed timestamps when interfacing it. I will work on checking this on all platforms using `timeutils`. However I don't have the machines available at hand to test my changes... What kind of integration test is run on baremetal platforms during CI tests ? I am looking at tests for `vfs_lfs`, `fatfs`, `modtime` and `machine_rtc`

yoctopuce commented:
I have pushed the new version, which should bring proper datetime support for most platforms.
There are a several platforms that still use an `uint64_t` when reading their RTC clock, but as the code setting the clock cannot use dates before 2000, this should not be an issue. When creating an mp_obj_t to hold a timestamp , they now use a proper function that will not fail when the timestamp is past a signed int.

yoctopuce commented:
> With which port did you test the PR?

I tested on ports/unix (using int64_t timestamps) and on our bare-metal port (an ARM platform using 1970 epoch, using 32 bit timestamps), but I don't have MicroPython classic bare metals available here. This is the reason I was asking if there was any possibility of running CI test remotely on bare metal hardware.

yoctopuce commented:
Anyone could suggest an idea to explein why the qemu_mips would be failing ?
The failing code is a new bit of the extra coverage written in C, which works properly in the standard `ports/unix`:
```c
        // convert a large timestamp value (stored in a mpz) to ll
        mp_printf(&mp_plat_print, "%lx\n", mp_obj_get_ll(obj_timestamp));

        // convert a small timestamp value (as a smallint) to a ll;
        obj_timestamp = timeutils_obj_from_timestamp(0xc0ffee);
        mp_printf(&mp_plat_print, "%lx\n", mp_obj_get_ll(obj_timestamp));
```
On qemu_mips, this is returning 0 instead of the proper values. The problem is likely the implementation of `mp_obj_get_ll()`, which uses `mp_obj_int_to_bytes_impl()` to retrieve the 64 bit value from the mpz number:
```c
int64_t res;
mp_obj_int_to_bytes_impl((mp_obj_t)arg, MP_ENDIANNESS_BIG, sizeof(res), (byte *)&res);
return res;
```
It is very similar to `ffifunc_print()` implementation, so I don't see why it would not work...

robert-hh commented:
I compiled this PR for a few boards, installed them on the board and ran the test script `time_mktime.py` with these boards. Result:

The test passes for STM32, ESP32 and MIMXRT
The test fails for CC3200, ESP8266, SAMD, Renesas-Ra, RP2 and NRF

The fail does not necessary tell that the change is wrong. Chances are better that it is a problem of the test script. Can you provide a few test cases and the expected result, which I can feed manually to the boards at REPL?

yoctopuce commented:
> I compiled this PR for a few boards, installed them on the board and ran the test script `time_mktime.py` with these boards. Result:
> 
> The test passes for STM32, ESP32 and MIMXRT The test fails for CC3200, ESP8266, SAMD, Renesas-Ra, RP2 and NRF
> 
> The fail does not necessary tell that the change is wrong. Chances are better that it is a problem of the test script. Can you provide a few test cases and the expected result, which I can feed manually to the boards at REPL?

Thank you very much, that helps a lot.

If you could try to run the script in `tests/extmod/time_mktime.py`, it should either report an error if I did make something really wrong, or display the properly supported date range for the platform. Knowing where it fails would problably give a fair amount of information on the issue.

If this test pass completely, maybe the problem is with another function.
You could for instance try
```py
import time
print(time.time())
```
to check if I broke something there.

robert-hh commented:
As suspected the problem is with running the test e.g. on RP2 with:
`./run-tests.py -t a0 extmod/time_mktime.py`

When I run the test with `mprmote run extmod/time_mktime.py` almost all ports return the line:
`tested from (1970, 1, 1) to (2100, 1, 1)`.

Exceptions:
NRF: `AttributeError: 'module' object has no attribute 'mktime'`
CC3200: is a little bit unreliable when executing code with `mpremote run`. Sometimes it fails.

Runtime of the test between 1.7s (Teensy 4.1) and 26.9s (Renesas-RA EK6M2).

yoctopuce commented:
> As suspected the problem is with running the test e.g. on RP2 with: `./run-tests.py -t a0 extmod/time_mktime.py`
> 
> When I run the test with `mprmote run extmod/time_mktime.py` almost all ports return the line: `tested from (1970, 1, 1) to (2100, 1, 1)`.
> 
> Exceptions: NRF: `AttributeError: 'module' object has no attribute 'mktime'` CC3200: is a little bit unreliable when executing code with `mpremote run`. Sometimes it fails.
> 
> Runtime of the test between 1.7s (Teensy 4.1) and 26.9s (Renesas-RA EK6M2).

Excellent, thank you very much for running these all these tests. So it looks like the objective of having an uniform supported date range is met.

The NRF error makes perfect sense: NRF port has very limited support for module `time`, it does not include `mktime`. I have just added a check at the beginning of my test to make sure to properly reports a `SKIP` rather than a failure when `time.mktime()` is unavailable.

Thank you again for you help testing this PR.

yoctopuce commented:
I did not anticipate that the test would take that long to run on bare metals. It might be a problem if it slows down the whole testing process.

I have just pushed a new version that skips faster over first days of the month until the 27th. This should run significantly faster. Can you give it a try ?

robert-hh commented:
All boards except cc3200 pass the test now when performed using run-tests.py. The script time_mktime.py itself runs fine at the CC3200 board. So it's a test run issue.  The problem seems to be in the code transport mechanism with pyboard.py. Even running the code with `mpremote run` is not reliable.
The CC3200 is already a little bit outdated. A fix of the test runner may not be needed.

yoctopuce commented:
> All boards except cc3200 pass the test now when performed using run-tests.py. The script time_mktime.py itself runs fine at the CC3200 board. So it's a test run issue. The problem seems to be in the code transport mechanism with pyboard.py. Even running the code with `mpremote run` is not reliable. The CC3200 is already a little bit outdated. A fix of the test runner may not be needed.

Excellent, thank you very much for your assistance !

While you have all the platforms available at hand, would you ave a chance to try to the other PR that I pushed yesterday, https://github.com/micropython/micropython/pull/17444 ? Same scenario as here: there is a new test file `tests/float/float_format_accuracy.py` that performs a 10'000 float conversions and tests the accuracy, but I now suspect my test code might be too slow to run on some platforms, as it was the case here.

robert-hh commented:
The slowest board in the previous test was the ESP8266. So I ran `run-tests.py -t u0 float/*.py`. It ran to an end, with the test summary below:

```
42 tests performed (2417 individual testcases)
34 tests passed
24 tests skipped: float/cmath_dunder.py float/cmath_fun.py float/cmath_fun_special.py float/complex1_intbig.py float/complex1.py float/complex_dunder.py float/complex_reverse_op.py float/complex_special_methods.py float/float2int_doubleprec_intbig.py float/float2int_intbig.py float/float_divmod.py float/float_format_ints_doubleprec.py float/float_format_ints_power10.py float/float_parse_doubleprec.py float/float_struct_e_doubleprec.py float/int_big_float.py float/math_constants_extra.py float/math_domain_special.py float/math_factorial_intbig.py float/math_fun_special.py float/math_isclose.py float/string_format.py float/true_value.py float/types.py
8 tests failed: float/builtin_float_round.py float/float1.py float/float_format_accuracy.py float/float_format_ints.py float/float_format.py float/float_parse.py float/lexer.py float/python36.py
```

Running the test on a SAMD21, only `float/float_format_accuracy.py` failed.
Renesas-RA EK-RA6M2 is slow as well, but all tests eithe pass or are skipped.

Besides that, the runtime should be fine.

yoctopuce commented:
> The tested ESP32, RP2 and NRF boards fail the `float_format_accuracy.py` test, because it takes too long. ESP8266 and SAM21 fail this test with:
> 
> ```
>   File "<stdin>", line 19, in pseudo_randouble
> AttributeError: 'module' object has no attribute 'crc32'
> ```

Thank you very much for the feedback, it looks like these platforms don't have built-in CRC32, I will put a Python implementation in my code. Would you mind copy-pasting your comments into the other PR so that things don't get mixes ?

Thanks again.

robert-hh commented:
I deleted my comments. You have to delete yours.

yoctopuce commented:
Excellent, thank you [robert-hh](https://github.com/robert-hh) for your help in testing this PR.
- We now have all automated CI checks passed
- Running `run-test` on STM32, ESP32, MIMXRT, ESP8266, SAMD, Renesas-Ra, RP2 and NRF is successful
- Running `time_mktime.py` manually on CC3200 is also successful
I guess we are ready for the next code review

dpgeorge commented on shared/timeutils/timeutils.h:
```diff
@@ -27,6 +27,46 @@
+//
+// By default this is enabled for machines using 64 bit pointers only,
+// but it can be enabled by specific ports
+#ifndef MP_SUPPORT_Y2100_AND_BEYOND
```
This is a configuration setting, so should start with `MICROPY_`.

I think it should be called `MICROPY_EPOCH_SUPPORT_Y2100_AND_BEYOND` and put in `py/mpconfig.h`, along with `MICROPY_EPOCH_IS_1970` and `MICROPY_EPOCH_IS_2000`.

yoctopuce replied:
Makes sense. I wasn't sure how much visibility you wanted to give to that option, but I am happy to put in `mpconfig.h`

dpgeorge commented on shared/timeutils/timeutils.h:
```diff
@@ -27,6 +27,46 @@
+#if MP_SUPPORT_Y2100_AND_BEYOND || MICROPY_EPOCH_IS_2000
+typedef int64_t platform_timestamp_t;
+#else
+typedef mp_uint_t platform_timestamp_t;
```
This type is used in `extmod` code so (due to code layering) I think it's better to define it in the core in `py/mpconfig.h`.

And also call it `mp_platform_timestamp_t` or even just `mp_timestamp_t` (that matches eg `mp_off_t`).

dpgeorge commented on shared/timeutils/timeutils.h:
```diff
@@ -27,6 +27,46 @@
+#if MP_SUPPORT_Y2100_AND_BEYOND
+typedef int64_t timeutils_timestamp_t;
+#else
+typedef mp_uint_t timeutils_timestamp_t;
```
This type is only used in `timeutils` so can stay named as is and defined here :+1:

dpgeorge commented on py/obj.c:
```diff
@@ -314,6 +314,36 @@ mp_int_t mp_obj_get_int(mp_const_obj_t arg) {
+    return mp_obj_int_get_uint_checked(arg);
+}
+
+int64_t mp_obj_get_ll(mp_const_obj_t arg) {
```
I would call this `mp_obj_get_int64()` because that's what it is.

Or, this function should return `long long` type.

yoctopuce replied:
I used `int64_t` because this is what other functions in `timeutils` were already using, and because I find it clear. However, from another perspective, there is an existing `mpz` function which is using the type `long long` to do the reverse operation (`mp_obj_new_int_from_ll()`), so it might make more sense to use `long long` as well.

Which option do you prefer ?

dpgeorge replied:
I suggest changing he type to `long long` and keeping the function name.

dpgeorge commented:
Running the tests on PYBD_SF6 everything passes except `tests/ports/stm32/modtime.py` which has the following diff for its output:
```diff
 time.localtime( -1 ) returned (1999, 12, 31, 23, 59, 59, 4, 365) (pass)
 time.localtime( 447549467 ) returned (2014, 3, 7, 23, 17, 47, 4, 66) (pass)
 time.localtime( -940984933 ) returned (1970, 3, 7, 23, 17, 47, 5, 66) (pass)
-time.localtime( -1072915199 ) returned (1966, 1, 1, 0, 0, 1, 5, 1) (pass)
-time.localtime( -1072915200 ) returned (1966, 1, 1, 0, 0, 0, 5, 1) (pass)
-time.localtime( -1072915201 ) returned (1965, 12, 31, 23, 59, 59, 4, 365) (pass)
+time.localtime( -1072915199 ) returned (2102, 2, 6, 6, 28, 17, 1, 37) expecting (1966, 1, 1, 0, 0, 1, 5, 1)
+time.localtime( -1072915200 ) returned (2102, 2, 6, 6, 28, 16, 1, 37) expecting (1966, 1, 1, 0, 0, 0, 5, 1)
+time.localtime( -1072915201 ) returned (2102, 2, 6, 6, 28, 15, 1, 37) expecting (1965, 12, 31, 23, 59, 59, 4, 365)
```
I'm not sure what to do here?  The new behaviour is probably more correct, but it's a breaking change.

yoctopuce commented:
> Running the tests on PYBD_SF6 everything passes except `tests/ports/stm32/modtime.py` which has the following diff for its output:
> 
> ```diff
>  time.localtime( -1 ) returned (1999, 12, 31, 23, 59, 59, 4, 365) (pass)
>  time.localtime( 447549467 ) returned (2014, 3, 7, 23, 17, 47, 4, 66) (pass)
>  time.localtime( -940984933 ) returned (1970, 3, 7, 23, 17, 47, 5, 66) (pass)
> -time.localtime( -1072915199 ) returned (1966, 1, 1, 0, 0, 1, 5, 1) (pass)
> -time.localtime( -1072915200 ) returned (1966, 1, 1, 0, 0, 0, 5, 1) (pass)
> -time.localtime( -1072915201 ) returned (1965, 12, 31, 23, 59, 59, 4, 365) (pass)
> +time.localtime( -1072915199 ) returned (2102, 2, 6, 6, 28, 17, 1, 37) expecting (1966, 1, 1, 0, 0, 1, 5, 1)
> +time.localtime( -1072915200 ) returned (2102, 2, 6, 6, 28, 16, 1, 37) expecting (1966, 1, 1, 0, 0, 0, 5, 1)
> +time.localtime( -1072915201 ) returned (2102, 2, 6, 6, 28, 15, 1, 37) expecting (1965, 12, 31, 23, 59, 59, 4, 365)
> ```
> 
> I'm not sure what to do here? The new behaviour is probably more correct, but it's a breaking change.

The proper solution is probably to add an extra setting `MP_SUPPORT_Y1969_AND_BEFORE`, that provides a similar extension of the supported timerange to older date. Given that the bit size of timestamps is already parametric, it should not be a big problem to add that.

dpgeorge commented on py/obj.c:
```diff
@@ -314,6 +314,36 @@ mp_int_t mp_obj_get_int(mp_const_obj_t arg) {
 }
 
+#if MICROPY_LONGINT_IMPL != MICROPY_LONGINT_IMPL_NONE
+mp_uint_t mp_obj_get_uint(mp_const_obj_t arg) {
```
A separate commit in this PR is fine.

dpgeorge commented:
> The proper solution is probably to add an extra setting `MP_SUPPORT_Y1969_AND_BEFORE`, that provides a similar extension of the supported timerange to older date.

OK, that sounds reasonable.  If you can keep that test passing, that would make things simpler.

yoctopuce committed:
py/obj: Add functions to retrieve large integers from mp_obj_t.

This commit provides helpers to retrieve integer values from
mp_obj_t when the content does not fit in a 32 bits integer,
without risking an implicit wrap due to an int overflow.

Signed-off-by: Yoctopuce dev <dev@yoctopuce.com>

yoctopuce committed:
shared/timeutils: Standardize supported date range on all platforms.

This is code makes sure that time functions work properly on a
reasonable date range, on all platforms, regardless of the epoch.
The suggested minimum range is 1970 to 2099.

In order to reduce code footprint, code to support far away dates
is only enabled specified by the port.

New types are defined to identify timestamps.

The implementation with the smallest code footprint is when
support timerange is limited to 1970-2099 and Epoch is 1970.
This makes it possible to use 32 bit unsigned integers for
all timestamps.

On ARM4F, adding support for dates up to year 3000 adds
460 bytes of code. Supporting dates back to 1600 adds
another 44 bytes of code.

Signed-off-by: Yoctopuce dev <dev@yoctopuce.com>

dpgeorge reviewed (commented):
This is looking very good now.

Tested on PYBD_SF6 and all tests now pass.

Also tested on ESP8266_GENERIC and all tests pass.

Just a few comments above to address.

yoctopuce commented on py/objint_longlong.c:
```diff
@@ -295,6 +295,22 @@ mp_int_t mp_obj_int_get_checked(mp_const_obj_t self_in) {
+            return truncated;
+        }
+    }
+    mp_raise_msg(&mp_type_OverflowError, MP_ERROR_TEXT("result overflows uint storage"));
```
Sure, done.

yoctopuce commented on py/mpconfig.h:
```diff
@@ -888,6 +888,64 @@ typedef double mp_float_t;
+// `timestamp_t` is the type that should be used by the port
+// to represent timestamps, and is referenced to the platform epoch
+#if MICROPY_TIMESTAMP_IMPL == MICROPY_TIMESTAMP_IMPL_LONG_LONG
+typedef long long timestamp_t;
```
Makes sense, I have changed it.

yoctopuce commented:
Rebased to master to fix the codecov report compared to master. Should be fine now.

dpgeorge reviewed (approved):
This is really nice improvement, thanks for your efforts here!

yoctopuce commented:
> This is really nice improvement, thanks for your efforts here!

You're welcome, thank you for helping to make it better !
